tempesta_fw/t/unit/test_tls.c

```diff
@@ -22,31 +22,419 @@
+ *	Testing mocks
+ * ------------------------------------------------------------------------
+ */
+void
+ttls_md_init(TlsMdCtx *ctx)
+{
+}
+
+void
+ttls_md_free(TlsMdCtx *ctx)
+{
+}
+
+int
+ttls_md_finish(TlsMdCtx *ctx, unsigned char *output)
+{
+	return 0;
+}
+
+int
+ttls_md(const TlsMdInfo *md_info, const unsigned char *input,
+		   size_t ilen, unsigned char *output)
+{
+	return 0;
+}
+
+int
+ttls_md_setup(TlsMdCtx *ctx, const TlsMdInfo *md_info, int hmac)
+{
+	return 0;
+}
+
+const TlsMdInfo *
+ttls_md_info_from_type(ttls_md_type_t md_type)
+{
+	return NULL;
+}
+
+int
+ttls_md_update(TlsMdCtx *ctx, const unsigned char *input, size_t ilen)
+{
+	return 0;
+}
+
+/*
+ * ------------------------------------------------------------------------
+ *	Tests for Tempesta TLS low-level basic math/crypto operations.
+ *
+ * Note that some of the routines are designed to be called in process
+ * context during Tempesta FW intialization and some of them are for
+ * run-time softirq context, so the testing routines must enable/disable
+ * FPU for them on its own.
+ * ------------------------------------------------------------------------
+ */
+#include "../../../tls/bignum.c"
+#include "../../../tls/ecp_curves.c"
+#include "../../../tls/ecp.c"
+#include "../../../tls/rsa_internal.c"
+#include "../../../tls/rsa.c"
+
+#define GCD_PAIR_COUNT  3
+
+static const int gcd_pairs[GCD_PAIR_COUNT][3] =
+{
+	{ 693, 609, 21 },
+	{ 1764, 868, 28 },
+	{ 768454923, 542167814, 1 }
+};
+
+/**
+ * MPI uses FPU only for primes, so no need to store/restore FPU state now.
+ * TODO #1064: add tests for primes.
+ */
+TEST(tls, mpi)
+{
+	int ret, i;
+	ttls_mpi A, E, N, X, Y, U, V;
+
+	ttls_mpi_init(&Y);
+	ttls_mpi_init(&V);
+	ttls_mpi_init(&A);
+	TTLS_MPI_CHK(ttls_mpi_read_string(&A, 16,
+					  "EFE021C2645FD1DC586E69184AF4A31E"
+					  "D5F53E93B5F123FA41680867BA110131"
+					  "944FE7952E2517337780CB0DB80E61AA"
+					  "E7C8DDC6C5C6AADEB34EB38A2F40D5E6"));
+	ttls_mpi_init(&E);
+	TTLS_MPI_CHK(ttls_mpi_read_string(&E, 16,
+					  "B2E7EFD37075B9F03FF989C7C5051C20"
+					  "34D2A323810251127E7BF8625A4F49A5"
+					  "F3E27F4DA8BD59C47D6DAABA4C8127BD"
+					  "5B5C25763222FEFCCFC38B832366C29E"));
+	ttls_mpi_init(&N);
+	TTLS_MPI_CHK(ttls_mpi_read_string(&N, 16,
+					  "0066A198186C18C10B2F5ED9B522752A"
+					  "9830B69916E535C8F047518A889A43A5"
+					  "94B6BED27A168D31D4A52F88925AA8F5"));
+	ttls_mpi_init(&X);
+	TTLS_MPI_CHK(ttls_mpi_mul_mpi(&X, &A, &N));
+	ttls_mpi_init(&U);
+	TTLS_MPI_CHK(ttls_mpi_read_string(&U, 16,
+					  "602AB7ECA597A3D6B56FF9829A5E8B85"
+					  "9E857EA95A03512E2BAE7391688D264A"
+					  "A5663B0341DB9CCFD2C4C5F421FEC814"
+					  "8001B72E848A38CAE1C65F78E56ABDEF"
+					  "E12D3C039B8A02D6BE593F0BBBDA56F1"
+					  "ECF677152EF804370C1A305CAF3B5BF1"
+					  "30879B56C61DE584A0F53A2447A51E"));
+
+	pr_info("  MPI test #1 (mul_mpi): ");
+
+	if ((ret = ttls_mpi_cmp_mpi(&X, &U)))
+		goto cleanup;
+	pr_info("passed\n");
```

Comment:
The `passed` messages used in mbedTLS are not informative, better to use the following approach to highlight where actually the fail has happened:
```
EXPECT_ZERO((ret = ttls_mpi_cmp_mpi(&X, &U)));
if (ret)
    goto cleanup;
````